src/global/right-click/heroes/dimensional-anchor.js

```diff
@@ -0,0 +1,261 @@
+    assert(card instanceof Card);
+    const cardName = ObjectNamespace.parseCard(card).name;
+    assert(cardName === CARD_NAME);
+
+    let cardOwnerSlot = -1;
+    for (const desk of world.TI4.getAllPlayerDesks()) {
+        const faction = world.TI4.getFactionByPlayerSlot(desk.playerSlot);
+        if (faction && faction.raw.leaders.heroes.includes(CARD_NAME)) {
+            cardOwnerSlot = desk.playerSlot;
+            break;
+        }
+    }
+
+    if (cardOwnerSlot === -1) {
+        Broadcast.broadcastAll(
+            locale("ui.message.dimensional_anchor_has_no_owner")
+        );
+        return;
+    }
+
+    const playerName = world.TI4.getNameByPlayerSlot(cardOwnerSlot);
+
+    Broadcast.broadcastAll(
+        locale("ui.message.dimensional_anchor", { playerName })
+    );
+
+    // Get all hexes with a system tile (much cheaper than getSystemTileObjectByPosition).
+    const hexSet = new Set();
+    for (const systemTileObj of world.TI4.getAllSystemTileObjects()) {
+        const pos = systemTileObj.getPosition();
+        const hex = Hex.fromPosition(pos);
+        hexSet.add(hex);
+    }
+
+    // get all systems with dimensional tears
+    const dimensionalTearHexes = new Set();
+    for (const obj of world.getAllObjects()) {
+        assert(obj instanceof GameObject);
+
+        const name = ObjectNamespace.parseGeneric(obj);
+
+        if (name && name.name === "dimensional_tear") {
+            const pos = obj.getPosition();
+            const hex = Hex.fromPosition(pos);
+            if (!hexSet.has(hex)) {
+                continue; // not on a system tile
+            }
+            dimensionalTearHexes.add(hex);
+        }
+    }
+
+    // get all systems adjacent to dimensional tears
+    const allHexes = new Set();
+    for (const hex of dimensionalTearHexes) {
+        allHexes.add(hex); // dimensional anchor applies to ships *in* or adjacent to
+
+        const adjNeighbor = new AdjacencyNeighbor(hex);
+        for (const adjHex of adjNeighbor.getAdjacent()) {
+            allHexes.add(adjHex);
+        }
+
+        const adjWormhole = new AdjacencyWormhole(hex, player.getSlot());
+        for (const adjHex of adjWormhole.getAdjacent()) {
+            allHexes.add(adjHex);
+        }
+
+        const adjHyperlane = new AdjacencyHyperlane(hex);
+        for (const adjHex of adjHyperlane.getAdjacent()) {
+            allHexes.add(adjHex);
+        }
+    }
+
+    // Remove any hexes that do not contain a system tile.
+    const trueHexes = new Set();
+    allHexes.forEach((hex) => {
+        if (hexSet.has(hex)) {
+            trueHexes.add(hex);
+        }
+    });
+
+    // get all non-fighter ships in those systems not owned by the player
+    // whose hero this is
+    const plastic = UnitPlastic.getAll().filter((plastic) =>
+        trueHexes.has(plastic.hex)
+    );
+
+    const ships = plastic.filter((plastic) => {
+        return (
+            SHIPS.includes(plastic.unit) &&
```

Comment:
Instead of hard-coding `SHIPS`, you could check the unitattrs:

```
const unitAttrsSet = new UnitAttrsSet() // default values
...
const isShip = unitAttrsSet.get(plastic.unit).raw.ship
const isNonFighterShip = isShip && plastic.unit !== 'fighter'
```

This ought to be resilient to adding new ship types (most likely homebrew).